PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx

```diff
@@ -135,32 +147,36 @@ struct HfTaskFlowCharmHadrons {
   /// \param cand is the candidate
   /// \param tracksQx is the X component of the Q vector for the tracks
   /// \param tracksQy is the Y component of the Q vector for the tracks
-  template <typename T1>
+  /// \param channel is the decay channel
+  template <int channel, typename T1>
   void getQvecDtracks(const T1& cand,
```

Comment:
```suggestion
  template <DecayChannel channel, typename T1>
  void getQvecDtracks(const T1& cand,
```

Reply:
Also remove `&` from `float& ampl`.

Reply:
@vkucera Hi, vit. You said that once. But when I change the `int` type to `DecayChannel` and build, there will be a failure.
`/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:152:8: note: candidate: 'template<DecayChannel channel, class T1> void HfTaskFlowCharmHadrons::getQvecDtracks(const T1&, std::vector<float>&, std::vector<float>&, float)'
  152 |   void getQvecDtracks(const T1& cand,
      |        ^~~~~~~~~~~~~~
/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:152:8: note:   template argument deduction/substitution failed:
/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:364:37: error: invalid conversion from 'int' to 'DecayChannel' [-fpermissive]
  364 |         getQvecDtracks<DecayChannel>(candidate, tracksQx, tracksQy, ampl);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                     |
      |                                     int
ninja: build stopped: subcommand failed.`
In my view, it is caused by the definition of `runFlowAnalysis`.
`  template <int DecayChannel, typename T1>
  void runFlowAnalysis(CollsWithQvecs::iterator const& collision,
                       T1 const& candidates)`

Reply:
Well, that's because you are not using `template<DecayChannel channel` in both `runFlowAnalysis` and in `getQvecDtracks` consistently.

Reply:
Yes, you are right. But the condition of `switch` must be int, can't be a class, namespace, or enumeration. And if the argument of template is `DecayChannel`, it should be used `if`, not `switch`. I'm asking Stefano's opinion.
For the first comment, they should be already sliced since the collision and the d meson tables possess the globalIndex.

Reply:
OK, thanks for your professional comment. I learned a lot.
The following is the case that used `template <DecayChannel Channel, typename T1>`
> `/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:303:16: error: 'Channel' is not a class, namespace, or enumeration
  303 |           case Channel::DsToKKPi:
      |                ^~~~~~~
/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:310:16: error: 'Channel' is not a class, namespace, or enumeration
  310 |           case Channel::DsToPiKK:
      |                ^~~~~~~
/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:329:16: error: 'Channel' is not a class, namespace, or enumeration
  329 |           case Channel::D0ToPiK:
      |                ^~~~~~~
/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:336:16: error: 'Channel' is not a class, namespace, or enumeration
  336 |           case Channel::D0ToKPi:
      |                ^~~~~~~
ninja: build stopped: subcommand failed.
`

Also, I tried to make the convertibility explicit with `enum DecayChannel : int {`. But I also got the error message. Based on the `taskCharmPolarisation.cxx`, I changed it to `enum DecayChannel : uint8_t {`, and got the same error. 

> `/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:286:8: note: candidate: 'template<<declaration error>, class T1>`

Meantime, the temple in `taskCharmPolarisation.cxx` is also used `init`.

> `  template <uint8_t channel, bool withMl, typename Cand>
  void runPolarisationAnalysis(Cand const& candidate, int bkgRotationId = 0)
  {`
  
  Could you please give me some samples for coding?

Reply:
OK. I know what you mean. I tried this modification. But I find the reason why occur `/home/wuct/alice/sw/SOURCES/O2Physics/master/0/PWGHF/D2H/Tasks/taskFlowCharmHadrons.cxx:286:8: note: candidate: 'template<<declaration error>, class T1>`
 is 
 `enum DecayChannel : int {` or `enum DecayChannel : uint8_t {`. The control variable method, which I've tried one by one, is where the problem lies.
 Thank you so much for your patience.

Reply:
I'm sorry, I don't understand the situation from your reply. Is it working now or is there still a problem. As I mentioned before, I think that you cannot use template argument as conditions of `switch` statements, so it that is the part that does not work, you have to use `if constexpr`.